tools/envoy_collect/envoy_collect.py

```diff
@@ -0,0 +1,243 @@
+    periodic).
+  - Validate in performance mode that we're using an opt binary.
+  - Consider handling other signals.
+  - bz2 compress tarball.
+  - Use freeze or something similar to build a static binary with embedded
+    Python, ending need to have Python on remote host (and care about version).
+"""
+from __future__ import print_function
+
+import argparse
+import ctypes
+import ctypes.util
+import datetime
+import json
+import os
+import pipes
+import shutil
+import signal
+import subprocess as sp
+import sys
+import tarfile
+import tempfile
+from six.moves import urllib
+
+DEFAULT_ENVOY_PATH = os.getenv('ENVOY_PATH',
+                               'bazel-bin/source/exe/envoy-static')
+PERF_PATH = os.getenv('PERF_PATH', 'perf')
+
+PR_SET_PDEATHSIG = 1  # See prtcl(2).
+
+DUMP_HANDLERS = ['clusters', 'listeners', 'server_info', 'stats']
+
+
+def FetchUrl(url):
+  return urllib.request.urlopen(url).read().decode('utf-8')
+
+
+def ModifyEnvoyconfig(config_path, perf, output_directory):
+  """Modify Envoy config to support gathering logs, etc.
+
+  Args:
+    config_path: the command-line specified Envoy config path.
+    perf: boolean indicating whether in performance mode.
+    output_directory: directory path for additional generated files.
+  Returns:
+    (modified Envoy config path, list of additional files to collect)
+  """
+
+  # No modifications yet when in performance profiling mode.
+  if perf:
+    return config_path, []
+
+  # Load original Envoy config.
+  with open(config_path, 'r') as f:
+    envoy_config = json.loads(f.read())
+
+  # Add unconditional access logs for all listeners.
+  access_log_paths = []
+  for n, listener in enumerate(envoy_config['listeners']):
+    for network_filter in listener['filters']:
+      if network_filter['name'] == 'http_connection_manager':
+        config = network_filter['config']
+        access_log_path = os.path.join(output_directory, 'access_%d.log' % n)
+        access_log_config = {'path': access_log_path}
+        if 'access_log' in config:
+          config['access_log'].append(access_log_config)
+        else:
+          config['access_log'] = [access_log_config]
+        access_log_paths.append(access_log_path)
+
+  # Write out modified Envoy config.
+  modified_envoy_config_path = os.path.join(output_directory, 'config.json')
+  with open(modified_envoy_config_path, 'w') as f:
+    f.write(json.dumps(envoy_config, indent=2))
+
+  return modified_envoy_config_path, access_log_paths
+
+
+def EnvoyCollect(parse_result, unknown_args):
+  """Run Envoy and collect its artifacts.
+
+  Args:
+    parse_result: Namespace object with envoy_collect.py's args.
+    unknown_args: list of remaining args to pass to Envoy binary.
+  """
+  # Are we in performance mode? Otherwise, debug.
+  perf = parse_result.performance
+  envoy_tmpdir = tempfile.mkdtemp()
+  return_code = 1  # Non-zero default return.
+  # Try and do stuff with envoy_tmpdir, rm -rf regardless of success/failure.
+  try:
```

Comment:
nit: might be better to split the inside of the try into a different function just to avoid the extra nesting in this giant function.